### DISCUSSION TITLE
react hook form with react table

### DISCUSSION THREAD

**Discussion by niralparmar** [react-hook-form/react-hook-form#1790] [Q&A]
Hi all,
how can i integrate one or more react table with react hook form.  my data model include array of complex object with multiple fields and nested object. 
i should be able to mark it as required if uses has not entered any data in table. i have been looking all over internet but i could find any help.

so i would appreciate if anybody could help me.

**Comment by bluebill1049** [comment]
do you have a codesandbox for the type of table you are mention above?

**Reply by niralparmar** [reply]
@bluebill1049  thanks for the quick response,
i dont have at the moment but i am trying to  create one but i just dont know how. 
but here is the sample model data i am trying to build form (actual one has mode nested objects)

`{
  name: "",
  description: "",
  localaddress: [
    {
      street1: "",
      street2: "",
      city: ""
    }
  ],
  foreignaddress: [
    {
      street1: "",
      street2: "",
      city: ""
    }
  ]
}`

idea here is, for address i am planning to use react table, there will be a button table footer which will open a popup with "street1","street2","city" fields. once user enters that data, it will gets inserted to react able and gets submitted also i have to show require fields as well if there are no address added. instead popup up i may go with inline editing where each input needs to get validate as well.

let me know if that help.

**Reply by bluebill1049** [reply]
as long as it's a form, it should work out the box, but if your table have virtual cells then you will have to use custom register. https://react-hook-form.com/advanced-usage#Workingwithvirtualizedlists

**Reply by niralparmar** [reply]
@bluebill1049 really appreciate your response. my apologies for delaying in providing codesanbox.
but here it is 
https://codesandbox.io/s/dreamy-wright-h3hvo

i tried to put comments as much as i could.
please let me know if you need more info.

**Reply by bluebill1049** [reply]
i think you mean `useFieldArray`

**Reply by niralparmar** [reply]
@bluebill1049  not sure how can i use "useFieldArray" , also we have complex  data table/grid (react-table based) and form must have it.

do you see any issue with my implementation?
appreciate your help.

**Reply by bluebill1049** [reply]
do you have a requirement to delete swap insert and etc?

**Reply by niralparmar** [reply]
@bluebill1049 delete yes but no replace.  in more complex scenario i have a table with inline editing with auto complete fields to add record. so i cant escape our table component

thanks once again

**Reply by bluebill1049** [reply]
with delete, then you should use `useFieldArray` to render your table cells.

**Reply by niralparmar** [reply]
thanks @bluebill1049 , but the question is, how can i use "react-table" or any data table with in react-hook-form?  the main reason of using react-table is, we have written complex business login side and if we use "fieldarray" not sure we can implement same complexity

you see in this mockup, grid is marked as required along with other fields and "+" and "-" button will add and remove the record.
![image](https://user-images.githubusercontent.com/4327556/83974044-2b6c3680-a89f-11ea-927f-19f1e0c0c02f.png)

from mockup it seems fieldarray is sufficient but its more complicated.
i really appreciate your help.

**Reply by bluebill1049** [reply]
i see. i will recommend you do a poc with react-table with react-hook-form first. i have very little Exp with react-table.

**Reply by niralparmar** [reply]
thanks @bluebill1049  that was the POC :) , since react table is uncontrolled component not sure how i can pass its reference to react hook form.

**Reply by bluebill1049** [reply]
as long as you can register all those inputs into hook form, that should be fine.

**Reply by bluebill1049** [reply]
actually why not using both useForm and useFieldArray? you can combine them right? for the table repeat section use FieldArray and the rest questions just use useForm.

**Reply by niralparmar** [reply]
our table is highly customize and it has inline editing which has all the business logic for eg, cascade dropdown, conditional rendering of fields,search, show/hide column etc. it will take a lot of time to re implement with useFieldArray.

so adding  existing table into form make more feasible. just like form is getting data from react-select, is it not possible to get data from react-table and assign it for field?
